### PR TITLE
Refactor Controller Config API and add Connector API

### PR DESCRIPTION
**Pull request by ShanBai6**
Change the utility method into a new utility class.
Add Connector API

### CODE REVIEW

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ApiResponseGenerator.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import static com.teradata.prestomanager.common.ExtendedStatus.MULTI_STATUS;
+import static javax.ws.rs.core.Response.Status.BAD_REQUEST;
+import static javax.ws.rs.core.Response.Status.INTERNAL_SERVER_ERROR;
+
+public final class ApiResponseGenerator
```
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ApiResponseGenerator.java]: I would call this something more indicative of its role as a call sender; it doesn't generate responses, it sends responses to some group of nodes.

Maybe `ApiCallForwarder` (though it isn't limited to forwarding) or `RequestMultiplexer` (or something less silly-sounding).

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ApiResponseGenerator.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+{
+    private static final Logger LOGGER = LogManager.getLogger(ApiResponseGenerator.class);
+    private static NodeSet nodeSet = NodeSet.getInstance();
+
+    private ApiResponseGenerator() {}
+
+    public static Response getResponseFromRequest(String scope, ApiRequester apiRequester, Collection<Integer> nodeId)
```
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ApiResponseGenerator.java]: This method could use a better name, but I can't think of anything particularly good, except maybe `multiplexApiCall`.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerConnectorAPI.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import java.util.List;
+
+import static com.teradata.prestomanager.controller.ApiResponseGenerator.getResponseFromRequest;
+
+@Path("/connector")
```
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerConnectorAPI.java]: Change to `"/connectors"`.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerConnectorAPI.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+        return getResponseFromRequest(scope, apiRequester, nodeId);
+    }
+
+    @POST
+    @Path("/{file}")
+    @Produces({MediaType.TEXT_PLAIN})
+    public Response setConnectorFileByURL(
+            @PathParam("file") String file,
+            String url,
```
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerConnectorAPI.java]: I would make this the first or last parameter.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerConnectorAPI.java (hunk 3 above)
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerConnectorAPI.java]: Change to `"/connectors"`.

**Comment by ShanBai6** [comment]
Fixed.

File: presto-manager-common/src/main/java/com/teradata/prestomanager/common/ApiRequester.java [hunk 5]
```diff
@@ -111,37 +112,37 @@ public ApiRequester build()
...
         {
-            uriBuilder.resolveTemplate(name, value);
+            uriBuilder.resolveTemplate(requireNonNull(name), requireNonNull(value));
             return this;
         }
 
         public Builder queryParam(String name, Object... values)
         {
-            uriBuilder.queryParam(name, values);
+            uriBuilder.queryParam(requireNonNull(name), requireNonNull(values));
```
Comment (sudarsanm94) [inline presto-manager-common/src/main/java/com/teradata/prestomanager/common/ApiRequester.java]: Here you should check whether there's at least one `values`

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerConnectorAPI.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+@Singleton
+public class ControllerConnectorAPI
+{
+    @GET
+    @Produces({MediaType.TEXT_PLAIN})
+    public Response getConnectors(@QueryParam("scope") String scope,
+            @QueryParam("nodeId") List<Integer> nodeId)
+    {
+        ApiRequester apiRequester = ApiRequester.builder(ControllerConnectorAPI.class)
+                .httpMethod(HttpMethod.GET)
```
Comment (sudarsanm94) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerConnectorAPI.java]: static import for `GET`

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerConnectorAPI.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+            @QueryParam("scope") String scope,
+            @QueryParam("nodeId") List<Integer> nodeId)
+    {
+        ApiRequester apiRequester = ApiRequester.builder(ControllerConnectorAPI.class)
+                .pathMethod("deleteConnectorFile")
+                .httpMethod(HttpMethod.DELETE)
+                .resolveTemplate("file", file)
+                .accept(MediaType.TEXT_PLAIN)
+                .build();
+
```
Comment (sudarsanm94) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerConnectorAPI.java]: As a note: Don't forget to add an API for updating a particular property in the config or connector file.

File: presto-manager-common/src/main/java/com/teradata/prestomanager/common/ApiRequester.java (hunk 5 above)
Comment (jirassimok) [inline presto-manager-common/src/main/java/com/teradata/prestomanager/common/ApiRequester.java]: You can have a query parameter without values.
It's not so useful in our current API, but it is a valid condition.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerConnectorAPI.java (hunk 7 above)
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerConnectorAPI.java]: We haven't implemented that for agents.

**Comment by ShanBai6** [comment]
Fixed & merged

**Pull request merged**